File: common-logger/src/main/java/io/appmetrica/analytics/logger/common/impl/LogMessageConstructor.java
package io.appmetrica.analytics.logger.common.impl;

import android.util.Log;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.util.Arrays;
import java.util.Locale;

public class LogMessageConstructor {

    private static final String TAG = "[LogMessageConstructor]";

    public String construct(@NonNull String tag, @Nullable String message, @Nullable Object... args) {
        return tag + " " + prepareMessage(message, args);
    }

    public String construct(
        @NonNull String tag,
        @Nullable Throwable throwable,
        @Nullable String message,
        @Nullable Object... args
    ) {
        final String nonNullMessage = (message == null ? "" : message) + "\n" + Log.getStackTraceString(throwable);
        return construct(tag, nonNullMessage, args);
    }

    private String prepareMessage(@Nullable String message, @Nullable Object... args) {
        String resultString;
        if (message == null) {
            resultString = "";
        } else if (args == null || args.length == 0) {
            resultString = message;
        } else {
            try {
                resultString = String.format(Locale.US, message, args);
            } catch (Throwable e) {
                resultString = "Attention!!! Invalid log format. See exception details above. Message: " +
                    message + "; arguments: " + Arrays.toString(args);
                Log.e(TAG, resultString, e);
            }
        }
        return String.format(
            Locale.US,
            "[%d-%s] %s",
            Thread.currentThread().getId(),
            Thread.currentThread().getName(),
            resultString
        );
    }
}

File: common-logger/src/main/java/io/appmetrica/analytics/logger/common/impl/LogMessageConstructor.kt
package io.appmetrica.analytics.logger.common.impl

import android.util.Log
import java.util.Locale

internal class LogMessageConstructor @JvmOverloads constructor(
    private val maskApiKeysInLog: Boolean = false,
) {

    private val tag = "[LogMessageConstructor]"

    fun construct(tag: String, message: String?, vararg args: Any?): String {
        return tag + " " + prepareMessage(message, *args)
    }

    fun construct(tag: String, throwable: Throwable?, message: String?, vararg args: Any?): String {
        val nonNullMessage = (message.orEmpty()) + "\n" + Log.getStackTraceString(throwable)
        return construct(tag, nonNullMessage, *args)
    }

    private fun prepareMessage(message: String?, vararg args: Any?): String {
        var resultString = when {
            message == null -> ""
            args.isEmpty() -> message
            else -> try {
                String.format(Locale.US, message, *args)
            } catch (e: Throwable) {
                val errorMessage = "Attention!!! Invalid log format. See exception details above. Message: " +
                    "$message; arguments: ${args.contentToString()}"
                val maskedErrorMessage =
                    if (maskApiKeysInLog) ApiKeyLogSanitizer.maskApiKeysInLog(errorMessage) else errorMessage
                Log.e(this.tag, maskedErrorMessage, e)
                maskedErrorMessage
            }
        }
        if (maskApiKeysInLog) {
            resultString = ApiKeyLogSanitizer.maskApiKeysInLog(resultString)
        }
        return String.format(
            Locale.US,
            "[%d-%s] %s",
            currentThreadIdForLog(),
            Thread.currentThread().name,
            resultString
        )
    }

    @Suppress("DEPRECATION")
    private fun currentThreadIdForLog(): Long = Thread.currentThread().id
}
